File: frontend/src/ask/intent.ts
import type { Kb } from '../kb/kb'
import { BODY_PARTS } from '../types'

/**
 * What kind of question this is, from its words alone: a medication question wants
 * pharmacogenomic markers and the current medication list, a lab question wants lab results, a
 * symptom question wants recent symptoms and measurements. Rules, not a model: every type says
 * which words triggered it, so the recommendation can explain itself. Pure and local.
 */
export type QuestionType = 'medication' | 'labs' | 'symptoms' | 'family' | 'doctor' | 'report' | 'genetics'

export interface Intent {
  type: QuestionType
  /** Distinct words or phrases in the question that point to this type, in question order. */
  signals: string[]
}

/**
 * A question word matches a stem of five or more letters it starts with ("prescribed" ← "prescri"),
 * or a shorter entry exactly ("mom" must not match "moment"). Entries with a space are phrases.
 */
const SIGNALS: Record<QuestionType, string[]> = {
  medication: [
    'medicat',
    'medicine',
    'drug',
    'drugs',
    'dose',
    'doses',
    'dosage',
    'pill',
    'pills',
    'tablet',
    'prescri',
    'side effect',
    'interaction',
    'statin',
    'antidepressant',
    'painkiller',
    'antibiotic',
    'supplement',
    'metaboli',
    'taking',
    'mg',
  ],
  labs: [
    'lab',
    'labs',
    'blood test',
    'test result',
    'result',
    'cholesterol',
    'ldl',
    'hdl',
    'triglycerid',
    'glucose',
    'hba1c',
    'ferritin',
    'vitamin',
    'iron',
    'tsh',
    'crp',
    'creatinin',
    'reference range',
    'level',
    'panel',
    'trend',
  ],
  symptoms: [
    'symptom',
    'pain',
    'painful',
    'ache',
    'aches',
    'aching',
    'hurt',
    'hurts',
    'fever',
    'headache',
    'migraine',
    'tired',
    'fatigue',
    'rash',
    'cough',
    'dizzy',
    'dizziness',
    'nause',
    'swell',
    'itch',
    'itchy',
    'sleep',
    'insomnia',
    'feel',
    'feeling',
    'sore',
    'stiff',
    'temperature',
    'blood pressure',
    'weight',
  ],
  family: [
    'inherit',
    'heredit',
    'parent',
    'mother',
    'father',
    'mom',
    'dad',
    'child',
    'son',
    'sons',
    'daughter',
    'sibling',
    'brother',
    'sister',
    'family',
    'carrier',
    'pass on',
    'passed on',
    'compare',
    'both of',
    'kids',
  ],
  doctor: [
    'doctor',
    'appointment',
    'visit',
    'clinician',
    'physician',
    'specialist',
    'gp',
    'cardiologist',
    'rheumatologist',
    'what to ask',
    'questions to ask',
    'prepare',
  ],
  report: [
    'report',
    'second opinion',
    'letter',
    'diagnos',
    'conclusion',
    'discharge',
    'scan',
    'scans',
    'mri',
    'x-ray',
    'ultrasound',
  ],
  genetics: [
    'gene',
    'genes',
    'genetic',
    'variant',
    'snp',
    'snps',
    'genotype',
    'dna',
    'mutation',
    'allele',
    'risk',
    'risks',
    'predispos',
  ],
}

/** Ties go to the more specific type. */
const ORDER: QuestionType[] = ['medication', 'labs', 'symptoms', 'report', 'doctor', 'family', 'genetics']

function words(text: string): string[] {
  return text
    .toLowerCase()
    .split(/[^\p{L}\p{N}*-]+/u)
    .filter(Boolean)
}

function hits(question: string, stems: string[]): string[] {
  const q = question.toLowerCase()
  const ws = words(question)
  const out: string[] = []
  for (const stem of stems) {
    const found = stem.includes(' ')
      ? q.includes(stem) && stem
      : ws.find((w) => w === stem || (stem.length >= 5 && w.startsWith(stem)))
    if (found && !out.includes(found)) out.push(found)
  }
  return out
}

/**
 * Every type the question shows signals for, strongest first. Knowledge-base names count too:
 * a drug from the kb is a medication signal, a gene symbol or rsid a genetics one, a body part a
 * symptom one. Asking about two or more selected people adds a family signal.
 */
export function classifyQuestion(question: string, kb: Kb, peopleSelected = 1): Intent[] {
  const ws = new Set(words(question))
  const extra: Record<QuestionType, string[]> = {
    medication: [],
    labs: [],
    symptoms: [],
    family: [],
    doctor: [],
    report: [],
    genetics: [],
  }
  for (const e of kb.entries) {
    for (const d of e.drugs ?? []) if (ws.has(d.toLowerCase())) extra.medication.push(d.toLowerCase())
    if (ws.has(e.gene.toLowerCase())) extra.genetics.push(e.gene)
    if (ws.has(e.rsid)) extra.genetics.push(e.rsid)
  }
  for (const w of ws) if (/^rs\d+$/.test(w) && !extra.genetics.includes(w)) extra.genetics.push(w)
  for (const b of BODY_PARTS) if (question.toLowerCase().includes(b)) extra.symptoms.push(b)
  if (peopleSelected >= 2) extra.family.push('several people')

  return ORDER.map((type) => ({
    type,
    signals: [...new Set([...hits(question, SIGNALS[type]), ...extra[type]])],
  }))
    .filter((i) => i.signals.length > 0)
    .sort((a, b) => b.signals.length - a.signals.length || ORDER.indexOf(a.type) - ORDER.indexOf(b.type))
}

/** The prompt template that fits a question type best (ids from prompts.ts). */
export const TEMPLATE_FOR: Record<QuestionType, string> = {
  medication: 'medication',
  labs: 'labs',
  symptoms: 'symptoms',
  family: 'compare',
  doctor: 'doctor',
  report: 'second-opinion',
  genetics: 'explain',
}

File: mobile/android/app/src/main/kotlin/com/hearth/ask/Intent.kt
package com.hearth.ask

import com.hearth.health.BODY_PARTS
import com.hearth.kb.Kb

/**
 * What kind of question this is, from its words alone (intent.ts): rules, not a model, and every
 * type says which words triggered it, so the recommendation can explain itself.
 */
enum class QuestionType(val id: String, val template: String) {
    MEDICATION("medication", "medication"),
    LABS("labs", "labs"),
    SYMPTOMS("symptoms", "symptoms"),
    FAMILY("family", "compare"),
    DOCTOR("doctor", "doctor"),
    REPORT("report", "second-opinion"),
    GENETICS("genetics", "explain"),
}

data class Intent(val type: QuestionType, val signals: List<String>)

/**
 * A question word matches a stem of five or more letters it starts with ("prescribed" ← "prescri"),
 * or a shorter entry exactly ("mom" must not match "moment"). Entries with a space are phrases.
 */
private val SIGNALS: Map<QuestionType, List<String>> = mapOf(
    QuestionType.MEDICATION to listOf(
        "medicat", "medicine", "drug", "drugs", "dose", "doses", "dosage", "pill", "pills", "tablet", "prescri", "side effect",
        "interaction", "statin", "antidepressant", "painkiller", "antibiotic", "supplement", "metaboli", "taking", "mg",
    ),
    QuestionType.LABS to listOf(
        "lab", "labs", "blood test", "test result", "result", "cholesterol", "ldl", "hdl", "triglycerid", "glucose", "hba1c",
        "ferritin", "vitamin", "iron", "tsh", "crp", "creatinin", "reference range", "level", "panel", "trend",
    ),
    QuestionType.SYMPTOMS to listOf(
        "symptom", "pain", "painful", "ache", "aches", "aching", "hurt", "hurts", "fever", "headache", "migraine", "tired",
        "fatigue", "rash", "cough", "dizzy", "dizziness", "nause", "swell", "itch", "itchy", "sleep", "insomnia", "feel",
        "feeling", "sore", "stiff", "temperature", "blood pressure", "weight",
    ),
    QuestionType.FAMILY to listOf(
        "inherit", "heredit", "parent", "mother", "father", "mom", "dad", "child", "son", "sons", "daughter", "sibling",
        "brother", "sister", "family", "carrier", "pass on", "passed on", "compare", "both of", "kids",
    ),
    QuestionType.DOCTOR to listOf(
        "doctor", "appointment", "visit", "clinician", "physician", "specialist", "gp", "cardiologist", "rheumatologist",
        "what to ask", "questions to ask", "prepare",
    ),
    QuestionType.REPORT to listOf(
        "report", "second opinion", "letter", "diagnos", "conclusion", "discharge", "scan", "scans", "mri", "x-ray", "ultrasound",
    ),
    QuestionType.GENETICS to listOf(
        "gene", "genes", "genetic", "variant", "snp", "snps", "genotype", "dna", "mutation", "allele", "risk", "risks", "predispos",
    ),
)

/** Ties go to the more specific type. */
private val ORDER = listOf(
    QuestionType.MEDICATION, QuestionType.LABS, QuestionType.SYMPTOMS, QuestionType.REPORT, QuestionType.DOCTOR,
    QuestionType.FAMILY, QuestionType.GENETICS,
)

private val NON_WORD = Regex("[^\\p{L}\\p{N}*-]+")
private val RSID = Regex("^rs\\d+$")

private fun words(text: String): List<String> = text.lowercase().split(NON_WORD).filter { it.isNotEmpty() }

private fun hits(question: String, stems: List<String>): List<String> {
    val q = question.lowercase()
    val ws = words(question)
    val out = mutableListOf<String>()
    for (stem in stems) {
        val found = if (' ' in stem) stem.takeIf { q.contains(it) } else ws.firstOrNull { it == stem || (stem.length >= 5 && it.startsWith(stem)) }
        if (found != null && found !in out) out.add(found)
    }
    return out
}

/**
 * Every type the question shows signals for, strongest first. Knowledge-base names count too: a
 * kb drug is a medication signal, a gene symbol or rsid a genetics one, a body part a symptom one.
 * Asking about two or more selected people adds a family signal.
 */
fun classifyQuestion(question: String, kb: Kb, peopleSelected: Int = 1): List<Intent> {
    val ws = words(question).toSet()
    val extra = QuestionType.entries.associateWith { mutableListOf<String>() }
    for (e in kb.entries) {
        for (d in e.drugs) if (d.lowercase() in ws) extra.getValue(QuestionType.MEDICATION).add(d.lowercase())
        if (e.gene.lowercase() in ws) extra.getValue(QuestionType.GENETICS).add(e.gene)
        if (e.rsid in ws) extra.getValue(QuestionType.GENETICS).add(e.rsid)
    }
    for (w in ws) if (RSID.matches(w) && w !in extra.getValue(QuestionType.GENETICS)) extra.getValue(QuestionType.GENETICS).add(w)
    for (b in BODY_PARTS) if (question.lowercase().contains(b)) extra.getValue(QuestionType.SYMPTOMS).add(b)
    if (peopleSelected >= 2) extra.getValue(QuestionType.FAMILY).add("several people")
    return ORDER.map { Intent(it, (hits(question, SIGNALS.getValue(it)) + extra.getValue(it)).distinct()) }
        .filter { it.signals.isNotEmpty() }
        .sortedWith(compareByDescending<Intent> { it.signals.size }.thenBy { ORDER.indexOf(it.type) })
}
